Update readme, add instructions for payment permit locally (#607)
ubiquity/ubiquibot: README.md

1. Create a new issue
2. Add a time label, ex: `Time: <1 Day`
3. Add a priority label, ex: `Priority: 0 (Normal)`
4. At this point the bot should add a price label.

## Configuration

`evm-network-id` is ID of the EVM-compatible network that will be used for payouts.

> `price-multiplier` is a base number that will be used to calculate bounty price based on the following formula: `price = price-multiplier * time-label-weight * priority-label-weight / 10`

Reviewer: Do we still have `/ 10`?

Author: It appears it is 100 now.

Even though looks like 1000.
https://github.com/ubiquity/ubiquibot/blob/d56a523c81829e17513f7da1f1ccb1bbdca8b180/src/handlers/shared/pricing.ts#L7

but priority is being being divided by 10 too. 
https://github.com/ubiquity/ubiquibot/blob/d56a523c81829e17513f7da1f1ccb1bbdca8b180/src/handlers/shared/pricing.ts#L6

> `price-multiplier` is a base number that will be used to calculate bounty price based on the following formula: `price = price-multiplier * time-label-weight * priority-label-weight / 100`

Reviewer: Is / 100 still accurate?

Author: oh that is multiply, not divide.

Author: I have tested this to make sure I am not wrong https://github.com/web4erOrg/ubiquibot/issues/58

- `name` is a human-readable name
- `weight` is a number that will be used to calculate the bounty price

`command-settings` are setting to enable or disable a command

- `name` is the name of the command
- `enabled` is a `true` or `false` value to enable or disable a command

`default-labels` are labels that are applied when an issue is created without any time or priority labels.

`assistive-pricing` to create a new pricing label if it doesn't exist. Can be `true` or `false`.

`disable-analytics` can be `true` or `false` that disables or enables weekly analytics collection by Ubiquity.

> `payment-permit-max-price` sets the amount for automatic payout of bounties when the issue is closed

Reviewer: ```suggestion
`payment-permit-max-price` sets the max amount for automatic payout of bounties when the issue is closed
```

> `comment-incentives` can be `true` or `false` that enable or disable comment incentives. These are comments in the issue by either the creator of the bounty or other users.

Reviewer: ```suggestion
`comment-incentives` can be `true` or `false` that enable or disable comment incentives. These are payments generated for comments in the issue by contributors, excluding the assignee.
```

> `issue-creator-multiplier` is a number that defines a base multiplier for calculating incentive reward for the creator of the issue.

Reviewer: "Incentive reward" Is redundant

> `comment-element-pricing` defines how much is a part of the comment worth. For example `text: 0.1` means that any text in the comment will be multiplied by 0.1

Reviewer: Will add 0.1

Not sure why my GitHub client isn't letting me suggest changes anymore

> `max-concurrent-assigns` is the maximum number of bounties that can be assigned to a bounty hunter at once. This excludes bounties with pending pull request reviews.

Reviewer: ```suggestion
`max-concurrent-assigns` is the maximum number of bounties that can be assigned to a bounty hunter at once. This excludes bounties with delayed or approved pull request reviews.
```

> `register-wallet-with-verification` can be `true` or `false`. If enabled, it requires a signed message to set wallet address. This prevents users from setting wallet address from centralized exchanges.

Reviewer: ```suggestion
`register-wallet-with-verification` can be `true` or `false`. If enabled, it requires a signed message to set wallet address. This prevents users from setting wallet address from centralized exchanges, which would make payments impossible to claim.
```

> `promotion-comment` is a comment that is appended to issue comment when a payment permit is generated.

Reviewer: ```suggestion
`promotion-comment` is a message that is appended to the payment permit comment.
```

1. Create a new private repository in your Github organization with name `ubiquibot-config`
2. Add your ubiquibot app to `ubiquibot-config` repository.
3. Create a file `.github/ubiquibot-config.yml` in it. Fill the file with contents from [this file](https://github.com/ubiquity/ubiquibot/blob/development/.github/ubiquibot-config.yml).
4. Go to https://pay.ubq.fi/keygen and generate X25519 public/private key pair. Fill private key of your wallet's address in `PLAIN_TEXT` field and click `Encrypt`.
5. Copy the `CIPHER_TEXT` and append it to your repo `ubiquibot-config/.github/ubiquibot-config.yml` as

   `private-key-encrypted: "PASTE_YOUR_CIPHER_TEXT_HERE"`

6. Copy the `X25519_PRIVATE_KEY` and append it in your local ubiquibot repository `.env` file as

   `X25519_PRIVATE_KEY=PASTE_YOUR_X25519_PRIVATE_KEY_HERE`

## How to QA any additions to the bot

Make sure you have your local instance of ubiquibot running.

1. Fork the ubiquibot repo and add your local instance of ubiquibot to the forked repository.

> 2. Enable Github action running on the forked repo and allow `issues` on the settings tab.

Reviewer 2: ```suggestion
2. Go to repository settings and turn on `issues`.
```

Author: I see these actions inform the hunter about the wrong commit message format and lint problems before they create PR preventing the need for a forced push later. Sounds like a part of QA to me.

Reviewer 2: Hmm okay I was just worried the bot will be triggered by actions but if the QA bot is not installed on the repo it will just fail, so it's fine

6. Link that QA issue to the pull request as indicated on the template before requesting a review

## Payments Permits in a local instance

For payment to work in your local instance, ubiquibot must be set up in a Github organization. It will not work for a ubiquibot instance set up in a personal account. Once, you have an ubiquibot instance working in an organization, follow the steps given below:

1. Create a new private repository in your Github organization with name `ubiquibot-config`
2. Add your ubiquibot app to `ubiquibot-config` repository.
3. Create a file `.github/ubiquibot-config.yml` in it. Fill the file with contents from [this file](https://github.com/ubiquity/ubiquibot/blob/development/.github/ubiquibot-config.yml).
4. Go to https://pay.ubq.fi/keygen and generate X25519 public/private key pair. Fill private key of your wallet's address in `PLAIN_TEXT` field and click `Encrypt`.
5. Copy the `CIPHER_TEXT` and append it to your repo `ubiquibot-config/.github/ubiquibot-config.yml` as

   `private-key-encrypted: "PASTE_YOUR_CIPHER_TEXT_HERE"`

6. Copy the `x25519_PRIVATE_KEY` and append it in your local ubiquibot repository `.env` file as

>    `x25519_PRIVATE_KEY=PASTE_YOUR_x25519_PRIVATE_KEY_HERE`

Reviewer 2: can you just change to capital letter X because env names are case-sensitive

Author: > Look good but I fear some new contributors might be confused because there are 2 ways to run the bot, as github action or locally.

I can fix this by adding following paragraph after the heading.

**You can perform a QA using your local instance of ubiquibot, or ubiquibot-qa app. The following steps describe how you can perform a QA using ubiquibot-qa.**

But do they need to change `X25519_PRIVATE_KEY` when they are using the QA bot? Encrypted private key is read from org config. And they don't have access to it, so they will not need to know this workflow key. What do you think?

Reviewer 2: `X25519_PRIVATE_KEY` for github actions is here https://github.com/ubiquity/ubiquibot/blob/d56a523c81829e17513f7da1f1ccb1bbdca8b180/.github/workflows/bot.yml#L81

The problem I see with QAing with github actions is that they will want to change the workflow variables to use their own instance of supabase and logger...which will need a commit that will show up in the pull request so when they finish work they need to revert changes so we don't merge them to our development branch. 

I think we should only write instructions for running the bot locally because that's the easiest way to debug and QA, at least in my experience.

Author: I like this. Fewer complications at starting stage. I have updated docs to QA using local instances. I am marking it resolved, but feel free to mark it unresolved if something needs a change.
